### PR TITLE
nim: fix deps, deprecate and patch old versions

### PR DESCRIPTION
**Pull request by Buldram**
* Fix runtime dependencies for produced binaries:
  * Add -rpaths pointing at dependencies to std wrapper modules
  * Set version constraints for OpenSSL
  * Added SQLite3 variant for <2.0
* Parallelize build with make
* Deprecate 1.0.10 due to CVEs
* Deprecate 1.9.3 as it's an old development version
* Backport patch for CVE-2021-21372 to <1.2.10/<1.4.4, CVE-2021-21374 to 1.4.2 and CVE-2021-46872 to 1.4.*

### CODE REVIEW

**Comment by Buldram** [comment]
@dialvarezs I'm requesting a review if you're willing

**Comment by dialvarezs** [comment]
Hi @Buldram, thank you for this contribution, especially for taking security into account. To be honest, these changes go beyond my knowledge of Nim, but they seem reasonable to me.

I installed the patched versions on Linux, and they work fine. The CI appears to be failing due to timeouts, so everything should be okay.

**Comment by dialvarezs** [comment]
@spackbot re-run pipeline

**Review by dialvarezs — APPROVED** [review]

File: var/spack/repos/builtin/packages/nim/httpclient_ssl_default-1.4.patch [hunk 1]
```diff
@@ -0,0 +1,185 @@
+From c7d090c418f7847bdfa1fd3bfed831470665c3e9 Mon Sep 17 00:00:00 2001
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/nim/httpclient_ssl_default-1.4.patch]: Is this a patch from https://github.com/nim-lang/Nim/commit/abac35e7437dd1ac2b3687dfa51de7f9d4b6e853?  If so, could it be retrieved using the url?  (See an example at starting at https://github.com/spack/spack/blob/dd8dff7872056161f9320a3b8c0329fbbef69d35/var/spack/repos/builtin/packages/dealii/package.py#L264.)

Comment (tldahlgren) [reply var/spack/repos/builtin/packages/nim/httpclient_ssl_default-1.4.patch]: Upstreamed patches are preferred over patch files when possible. We already have issues with using too many inodes at some facilities. For more on patches, see https://spack.readthedocs.io/en/latest/packaging_guide.html#patch.

File: var/spack/repos/builtin/packages/nim/package.py [hunk 2]
```diff
@@ -18,39 +21,58 @@ class Nim(Package):
...
+        version(
+            "0.19.9",
+            sha256="154c440cb8f27da20b3d6b1a8cc03f63305073fb995bbf26ec9bc6ad891ce276",
+            url="https://github.com/nim-lang/nightlies/releases/download/2019-06-02-devel-1255b3c/nim-0.19.9-linux_x64.tar.xz",
+        )
+        version(
+            "0.19.6", sha256="a09f0c58d29392434d4fd6d15d4059cf7e013ae948413cb9233b8233d67e3a29"
+        )
+
+    variant("sqlite", default=False, when="@0:1.9", description="Install SQLite for std/db_sqlite")
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/nim/package.py]: Does this mean there is no `sqlite` variant available for versions > `1.9.z`?

**Review by tldahlgren — CHANGES_REQUESTED** [review]
Any chance at least some of the patch files can be upstreamed (i.e., retrieved from repository URLs)?  I also have a question about the `sqlite` variant change.

File: var/spack/repos/builtin/packages/nim/package.py (hunk 2 above)
Comment (Buldram) [reply var/spack/repos/builtin/packages/nim/package.py]: That's correct, it was moved into a package for 2.0
https://github.com/nim-lang/Nim/pull/20986

File: var/spack/repos/builtin/packages/nim/httpclient_ssl_default-1.4.patch (hunk 1 above)
Comment (Buldram) [reply var/spack/repos/builtin/packages/nim/httpclient_ssl_default-1.4.patch]: Done, thanks.

**Comment by Buldram** [comment]
@spackbot fix style

**Review by tldahlgren — APPROVED** [review]

**Comment by tldahlgren** [comment]
Thanks for the updates.

**Pull request merged**